cores/nRF5/Tone.cpp

```diff
@@ -75,118 +102,67 @@ void tone(uint8_t pin, unsigned int frequency, unsigned long duration)
 								0,
 								0
     };
-	
-#if 0
-	//assign pin to pwm channel - look at WVariant.h for details about ulPWMChannel attribute
-	uint8_t pwm_type=g_APinDescription[pin].ulPWMChannel;
-	if(pwm_type == NOT_ON_PWM)
-		return;
-	
-	uint32_t pins[NRF_PWM_CHANNEL_COUNT]={NRF_PWM_PIN_NOT_CONNECTED, NRF_PWM_PIN_NOT_CONNECTED, NRF_PWM_PIN_NOT_CONNECTED, NRF_PWM_PIN_NOT_CONNECTED};
-	pins[pwm_type & 0x0F]=g_APinDescription[pin].ulPin;
-	IRQn_Type IntNo = PWM0_IRQn;
-	NRF_PWM_Type * PWMInstance = NRF_PWM0;
-	switch(pwm_type &0xF0){
-		case 16://0x10
-			PWMInstance = NRF_PWM1;
-			IntNo = PWM1_IRQn;
-			break;
-		case 32://0x20
-			PWMInstance = NRF_PWM2;
-			IntNo = PWM2_IRQn;
-			break;
-	}
-#else
-	// Use fixed PWM2, TODO could conflict with other usage
+
 	uint32_t pins[NRF_PWM_CHANNEL_COUNT]={NRF_PWM_PIN_NOT_CONNECTED, NRF_PWM_PIN_NOT_CONNECTED, NRF_PWM_PIN_NOT_CONNECTED, NRF_PWM_PIN_NOT_CONNECTED};
 	pins[0] = g_ADigitalPinMap[pin];
 
-	IRQn_Type IntNo = PWM2_IRQn;
-	NRF_PWM_Type * PWMInstance = NRF_PWM2;
-#endif
-
-	nrf_pwm_pins_set(PWMInstance, pins);
-	nrf_pwm_enable(PWMInstance);
-	nrf_pwm_configure(PWMInstance, NRF_PWM_CLK_125kHz, NRF_PWM_MODE_UP, time_per);
-	nrf_pwm_decoder_set(PWMInstance, NRF_PWM_LOAD_COMMON, NRF_PWM_STEP_AUTO);
-	nrf_pwm_sequence_set(PWMInstance, 0, &seq);
-	nrf_pwm_shorts_enable(PWMInstance, NRF_PWM_SHORT_SEQEND0_STOP_MASK);
-	
 	// enable interrupt
-	nrf_pwm_event_clear(PWMInstance, NRF_PWM_EVENT_PWMPERIODEND);
-	nrf_pwm_int_enable(PWMInstance, NRF_PWM_INT_PWMPERIODEND_MASK);
-	NVIC_SetPriority(IntNo, 6); //low priority
-	NVIC_ClearPendingIRQ(IntNo);
-	NVIC_EnableIRQ(IntNo);
-
-	nrf_pwm_task_trigger(PWMInstance, NRF_PWM_TASK_SEQSTART0);
+	count_duration = 0x6FFF; // large enough to avoid hitting zero in next few lines
+	no_stop = new_no_stop;
+
+	nrf_pwm_pins_set(_PWMInstance, pins);
+	nrf_pwm_enable(_PWMInstance);
+	nrf_pwm_configure(_PWMInstance, NRF_PWM_CLK_125kHz, NRF_PWM_MODE_UP, time_per);
+	nrf_pwm_decoder_set(_PWMInstance, NRF_PWM_LOAD_COMMON, NRF_PWM_STEP_AUTO);
+	nrf_pwm_sequence_set(_PWMInstance, 0, &seq);
+	nrf_pwm_shorts_enable(_PWMInstance, NRF_PWM_SHORT_SEQEND0_STOP_MASK); // shortcut for when SEQ0 ends, PWM output will automatically stop
+	nrf_pwm_event_clear(_PWMInstance, NRF_PWM_EVENT_PWMPERIODEND);
+	nrf_pwm_int_enable(_PWMInstance, NRF_PWM_INT_PWMPERIODEND_MASK);
+	NVIC_SetPriority(_IntNo, 6); //low priority
+	NVIC_ClearPendingIRQ(_IntNo);
+	NVIC_EnableIRQ(_IntNo);
+	count_duration = new_count_duration;
```

Comment:
<details><summary>It's prevents an error, because prior call to `tone()` is not guaranteed to be finished.</summary><P/>

Specifically, this avoids a timing window where the device gets into a bad state, when the remaining count in the `volatile` variable `count_duration` is decremented to zero in the midst of this code block configuring and starting the PWM peripheral.

This code was written to mimic as closely as possible the previous code's interaction with the peripheral, without disabling interrupts, and has been functionally tested (see test gist).
</details>
<details><summary>Feel free to open an issue to investigate ...</summary><P/>

Feel free to open an issue to track looking into whether this similar affect is possible by calling `NVIC_DisableIRQ(PWM2_IRQn)`.  Unfortunately, I think much more thought would be needed to ensure that works correctly, such as when there is high radio use (long running SoftDevice interrupts that cannot be disabled).

I am not yet comfortable saying that such code would be solid.
</details>

Reply:
Yeah, I think you have overworked, API() requires user to call noTone() before calling tone() so yeah, it is simpler to just disable current IRQ and completely discard old duration to use the new configure.

Reply:
I understand the [Arduino docs](https://www.arduino.cc/reference/en/language/functions/advanced-io/tone/) differently. 
 I think they allow calling `tone()` to change (at least) the frequency for a pin that is already generating a tone.  Only when changing the pin is it required to first call `noTone()`.

Is there a different API you are referring to?

Reply:
yeah that is the same one, calling the tone() on same pin change the freq immediately, therefore we can just disable the IRQ within tone() and reset everything.